crates/bin/node/src/routes/check_state/mod.rs

```diff
@@ -0,0 +1,54 @@
+use nuts::nut01::PublicKey;
+use tonic::Status;
+
+use crate::grpc_service::GrpcState;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    // Database errors
+    #[error("failed to acquire database connection: {0}")]
+    DbConnection(#[source] sqlx::Error),
+    #[error("failed to retrieve proof state: {0}")]
+    ProofStateRetrieval(#[source] sqlx::Error),
+    #[error("invalid public key: {0}")]
+    InvalidPublicKey(#[source] nuts::nut01::Error),
+}
+
+impl From<Error> for Status {
+    fn from(error: Error) -> Self {
+        match error {
+            Error::DbConnection(_) => Status::internal("Database connection error"),
+            Error::ProofStateRetrieval(_) => Status::not_found("Failed to retrieve proof state"),
+            Error::InvalidPublicKey(_) => Status::invalid_argument("Invalid public key provided"),
+        }
+    }
+}
+impl GrpcState {
+    pub async fn inner_check_state(
+        &self,
+        ys: Vec<Vec<u8>>,
+    ) -> Result<nuts::nut07::CheckStateResponse, Error> {
+        let mut conn = self.pg_pool.acquire().await.map_err(Error::DbConnection)?;
+
+        let proof_data = db_node::proof::get_proofs_by_ids(&mut conn, &ys)
+            .await
+            .unwrap();
+
+        let proof_states: Result<Vec<_>, _> = ys
+            .iter()
+            .map(|y| {
+                let y = PublicKey::from_slice(y).map_err(Error::InvalidPublicKey)?;
+                let state = proof_data
+                    .get(y.to_bytes().as_slice())
+                    .unwrap_or(&nuts::nut07::ProofState::Unspent)
+                    .clone();
+
+                Ok(nuts::nut07::ProofCheckState { y, state })
+            })
+            .collect();
```

Comment:
Those two calls can be merged in a single one using sql `JOIN`  feature.

Reply:
Also you did not do the logic to make sure that the proofs are returned in the exact same orders as the `ys` at the database level.